File: messenger-server/messenger-server/src/main/java/dev/crashbandicootfm/messenger/service/service/chat/ChatService.java
package dev.crashbandicootfm.messenger.service.service.chat;

import dev.crashbandicootfm.messenger.service.exception.user.ChatException;
import dev.crashbandicootfm.messenger.service.exception.user.UserException;
import dev.crashbandicootfm.messenger.service.model.ChatModel;
import dev.crashbandicootfm.messenger.service.model.MessageModel;
import java.util.List;
import org.jetbrains.annotations.NotNull;

public interface ChatService {

  @NotNull ChatModel getById(Long id) throws ChatException;

  @NotNull ChatModel save(@NotNull ChatModel chatModel);

  @NotNull ChatModel createChat(@NotNull String chatName, @NotNull Long userId) throws ChatException, UserException;

  @NotNull ChatModel findByChatName(@NotNull String chatName) throws ChatException;

  @NotNull ChatModel joinChat(@NotNull String chatName, @NotNull Long userId) throws ChatException, UserException;

  @NotNull ChatModel createChatWithPassword(@NotNull String chatName, @NotNull String password) throws ChatException;

  @NotNull ChatModel joinChatWithPassword(@NotNull String chatName, @NotNull Long userId, @NotNull String password) throws ChatException, UserException;

  @NotNull List<ChatModel> findAllByUserId(@NotNull Long userId) throws UserException;

  @NotNull ChatModel startPrivateChat(@NotNull Long userId, @NotNull Long otherUserId) throws UserException, ChatException;

  void deleteChat(@NotNull Long chatId, @NotNull Long userId) throws ChatException;

  void leaveChat(@NotNull Long chatId, @NotNull Long userId) throws ChatException;

  @NotNull List<MessageModel> getMessagesByChatId(Long chatId);

  void markChatAsRead(Long chatId, Long userId, Long lastMessageId);

  Long getUnreadMessagesCount(Long chatId, Long userId);

  @NotNull List<ChatModel> getAllChatsByUserId(Long userId);
}

File: messenger-web-client/src/app/services/chat.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpHeaders} from '@angular/common/http';
import {Observable, throwError} from 'rxjs';
import {ChatRequest} from '../models/request/chat-request.model';
import {ChatResponse} from '../models/response/chat-response.model';
import {ChatWithPasswordRequest} from '../models/request/chat-with-password-request.model';
import {switchMap} from 'rxjs/operators';

@Injectable({
  providedIn: 'root'
})
export class ChatService {

  private readonly url = "http://localhost:8080/api/v1/chats/";

  constructor(private http: HttpClient) {}

  private getHeaders(): HttpHeaders {
    const token = localStorage.getItem('token');
    console.log("Token from service:", token);
    return new HttpHeaders({
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    });
  }

  createChat(name: string): Observable<ChatResponse> {
    const headers = this.getHeaders();
    const payload: ChatRequest = { name };

    return this.http.post<ChatResponse>(this.url, payload, { headers });
  }

  joinChat(name: string): Observable<ChatResponse> {
    const token = localStorage.getItem('token');
    console.log("Token from service:", token);
    const headers = new HttpHeaders({
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    });

    const payload = { name };

    return this.http.post<ChatResponse>(`${this.url}join`, payload, { headers });
  }

  createChatWithPassword(name: string, password: string): Observable<ChatResponse> {
    const headers = this.getHeaders();
    const payload: ChatWithPasswordRequest = { name, password };
    return this.http.post<ChatResponse>(`${this.url}createWithPassword`, payload, { headers });
  }

  joinChatWithPassword(name: string, password: string): Observable<ChatResponse> {
    const headers = this.getHeaders();
    const payload = { name };
    return this.http.post<ChatResponse>(`${this.url}join-with-password?password=${password}`, payload, { headers });
  }

  getUserChats(): Observable<ChatResponse[]> {
    const headers = this.getHeaders();
    return this.http.get<ChatResponse[]>(`${this.url}user/chats?timestamp=${Date.now()}`, { headers });
  }

  createPrivateChat(otherUserId: number): Observable<ChatResponse> {
    const headers = this.getHeaders();
    return this.http.post<ChatResponse>(`${this.url}start-private?otherUserId=${otherUserId}`, {}, { headers });
  }

  deleteChat(chatId: number): Observable<any> {
    const headers = this.getHeaders();
    return this.http.delete(`${this.url}${chatId}`, { headers });
  }

  leaveChat(chatId: number | null): Observable<any> {
    if (chatId === null) {
      return throwError(() => new Error('Chat ID is null'));
    }

    const headers = this.getHeaders();
    return this.http.post(`${this.url}leave?chatId=${chatId}`, {}, { headers });
  }

  getMessagesByChatId(chatId: number): Observable<any> {
    const headers = this.getHeaders();
    return this.http.get<any>(`${this.url}${chatId}/messages`, { headers });
  }

  getUnreadMessagesCount(chatId: number, userId: number): Observable<number> {
    return this.http.get<number>(`${this.url}unread-count?chatId=${chatId}&userId=${userId}`);
  }

  markChatAsRead(chatId: number, userId: number | null): Observable<void> {
    return this.http.post<void>(`${this.url}${chatId}/mark-as-read`, { userId });
  }
}
